src/main/java/org/passay/entropy/RandomPasswordEntropyFactory.java

```diff
@@ -0,0 +1,75 @@
+/* See LICENSE for licensing and NOTICE for copyright. */
+package org.passay.entropy;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.passay.AllowedCharacterRule;
+import org.passay.CharacterCharacteristicsRule;
+import org.passay.PasswordData;
+import org.passay.Rule;
+
+/**
+ * Factory for creating {@link RandomPasswordEntropy} from password rules and password data.
+ *
+ * @see {@link RandomPasswordEntropy}.
+ *
+ * @author  Middleware Services
+ */
+public final class RandomPasswordEntropyFactory
+{
+
+
+  /**
+   * Private constructor for factory class.
+   */
+  private RandomPasswordEntropyFactory() {}
+
+
+  /**
+   * Creates a new random password entropy.
+   *
+   * @param  passwordRules  to aid in entropy calculation
+   * @param  passwordData  to aid in entropy calculation
+   *
+   * @return  random password entropy
+   */
+  public static RandomPasswordEntropy createEntropy(final List<Rule> passwordRules, final PasswordData passwordData)
+  {
+    final Set<Character> uniqueCharacters = new HashSet<>();
+    passwordRules.stream().forEach((rule) -> {
+      if (rule instanceof CharacterCharacteristicsRule) {
+        final CharacterCharacteristicsRule characteristicRule = (CharacterCharacteristicsRule) rule;
+        characteristicRule.getRules().forEach((characterRule) ->
+          uniqueCharacters.addAll(getUniqueCharacters(characterRule.getValidCharacters())));
+      } else if (rule instanceof AllowedCharacterRule) {
```

Comment:
What of having a list of CharacterRule only?
```
    final List<Rule> l = new ArrayList<>();
    l.add(new CharacterRule(EnglishCharacterData.Digit, 1));
```
That was my reason for implementing CharacterData for the two classes and iterate with optimal big O
```
passwordRules.stream().forEach((rule) -> {
      if (rule instanceof CharacterCharacteristicsRule) {
        final CharacterCharacteristicsRule characteristicRule = (CharacterCharacteristicsRule) rule;
        characteristicRule.getRules().forEach((characterRule) -> {
     //  ...
        });
      } else if (implementsInterface(rule, CharacterData.class)) {
        final CharacterData characterData = (CharacterData) rule;
     //  ...
      } else if (AbstractDictionaryRule.class.isAssignableFrom(rule.getClass())) {
       //  ...
      }
    });
```
It's a valid scenario for PasswordValidator, would you agree?